### DISCUSSION TITLE
Using concurrency in workflows shouldn't cancel previously pending workflows

### DISCUSSION THREAD

**Discussion by TSRBerry** [community/community#41518] [Actions] [answered]
Better discussion:

https://github.com/orgs/community/discussions/12835

### Select Topic Area

Product Feedback, Question

### Body

We recently introduced concurrency for our [release workflow](https://github.com/Ryujinx/Ryujinx/actions/workflows/release.yml) at Ryujinx , so we could merge PRs without waiting for the previous release job to finish.

We expected all new workflows to be pending until the previous job is done, but that wasn't the case.

![image](https://user-images.githubusercontent.com/20988865/207078444-9a782189-2fd7-470b-8f62-d3e41636f8ff.png)

As you can see here, the first job gets canceled after a new pending job gets added, which we wanted to avoid.

![image](https://user-images.githubusercontent.com/20988865/207078737-2db776cd-891c-4469-99fc-92387dae8ba5.png)

We never defined a "priority" or anything to make it cancel this workflow, but it still did.

Reading the documentation again we found this sentence:

> Any previously pending job or workflow in the concurrency group will be canceled.

Source: [Github workflow concurrency](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency)

Now this job was already in progress, but was canceled anyway, so we aren't sure if this note applies to our situation.
`cancel-in-progress: true` is obviously not defined in our workflow file, so this situation is a little confusing.

If this is intended is there a way for us to not cancel previously in-progress workflows? We would prefer to run these workflows in the order they were queued in, so our changelog matches with the actual releases we publish.

---

(cc: @marysaka)

**Comment by TSRBerry** [comment]
(Seems like this discussion is missing a few labels, but I'm not sure what to do to add them.)

**Comment by mgasner** [comment]
See also: https://github.com/actions/runner/issues/1523, https://github.com/actions/runner/issues/2227 (folks from GH saying not their problem), https://github.com/orgs/community/discussions/5435 (folks from GH saying they won't implement, a year ago), https://github.com/orgs/community/discussions/33142, https://github.com/orgs/community/discussions/12835, https://github.com/orgs/community/discussions/33948, https://github.com/orgs/community/discussions/32376. Related: https://github.com/orgs/community/discussions/31040, https://github.com/orgs/community/discussions/31544. Plus lots of others, plus others discussing the need for configurable concurrency limits (e.g. limit to 2 concurrent jobs).

**Comment by vitalykarasik** [comment]
+1 - 'concurrency' implementation lacks a few important features.
Unfortunately, I don't know how we can raise an issue about that.
Vitaly
https://github.com/community/community/discussions/46014
https://www.linkedin.com/posts/vkarasik_cicd-jenkins-githubactions-activity-7026518468577366017-X0tt

**Comment by ShutdownRepo** [comment]
Agreed, this is a very important lack. I find it very surprising that GitHub's pipeline lacks the ability to manage queues

**Comment by austinpray-mixpanel** [comment]
My use case: I've got expensive e2e tests that I only want to run one master copy of at a time to leave headroom for all the PRs that might need the resources. I want to run on every master commit so we can easily bisect when they break. We have multiple commits per hour during working hours.

**Comment by yardensade** [comment]
How come pending jobs with concurrency are being cancelled? most of us will want all of them to await running and not be cancelled by new triggers.

This feature is very lacking as is and missing a very strong use case for customers, can we get some insight on this?

**Comment by SamBernaerdtOntoforce** [comment]
Same here. We need the option
`cancel-pending: false`

Our use case: We have e2e tests that run against the same server and cannot run in parallel. Any e2e workflow should only start when no other e2e workflows are running. Concurrency would be the solution if there is an option to *not* cancel pending jobs.

**Comment by MerzoukeMansouri** [comment]
+1 need a waitting list

**Comment by dvasdekis** [comment]
Mature CICD solutions (Azure Devops, CircleCI) have this solution, see here for Devops implementation: https://learn.microsoft.com/en-us/azure/devops/release-notes/2021/sprint-190-update

+1 to having `cancel-pending: false`. The expected behaviour is to queue.

**Comment by jlongo-coforge** [comment]
Any updates on this one?

**Comment by aquierick** [comment]
+1 to having cancel-pending: false

**Comment by iamironz** [comment]
+ for it. 
It would be way more convenient to support workflow runs one-by-one based on the queue instead of canceling, as it might be perhaps important for validating develop after the merge and sending Slack notifications to the appropriate contributor in the case of the failed build. We have exactly this case.

**Comment by groeney** [comment]
Does anyone have a clever workaround to not having support for `cancel-pending: false`?

**Reply by samirtahir91** [reply]
I have worked around by checking the status of last 50 jobs/runs for workflow ID(s), for all those without the status "completed", check the created_at timestamp, and essentially sleep and loop until all of the non-completed runs with an older timestamp are completed. The caveat here is your queued jobs are sleeping and using compute time.

I'm not using the concurrency feature at all now

In my case I have a workflow that does some changes, creates a PR and merges to main. The reason I have to queue is because Github API rate limiting, there is the option for merge queues however even creating a PR is limited by Githubs secondary api rate limit for a single user/account, basically you cannot do any parallel api calls to Github and their documentation strongly advises against it. Which is why i find it bizarre there is no native queuing option/serial run option.

It works ok for small time consuming jobs, but for longer jobs you may not want them to be sleeping for a long time

**Comment by vmred** [comment]
+1 to having cancel-pending: false or any other naming flag for this

**Comment by danlehmann** [comment]
+1 for "cancel-pending: false"

This feature would be incredibly useful for us as our automated test infra pretty often skips over tests when using concurrency. While we can come up with a way to enforce concurrency inside the tests themselves, that is pretty involved (and has the risk of running into timeouts). Github's feature is so close to being really useful

**Comment by eliasdefaria** [comment]
+1 for "cancel-pending: false"

**Comment by harpener** [comment]
Hi guys, our use case is that we need a passive queue mechanism, two levels of concurrency instead of just one. Our workflow for status checks should cancel pending tasks from the same group 1 (pull request), but keep pending tasks from a different group 2 (repository). Essentially, the workflow should never run in parallel, but should always be queued if it comes from a different pull request. Not being able to specify more groups would be fine as long as we can at least disable cancelling pending tasks at all. This is what we had in Jenkins CI. I find it hard to believe that this is such a rare use case unsolved for years.

**Comment by starikcetin** [comment]
This is absolutely needed.

**Comment by guyeisenbach** [comment]
+1 for "cancel-pending: false"

**Comment by DanBezalelpx** [comment]
Agree. Much needed. would be great!

**Comment by paperwriter0300** [comment]
Absolutely, implementing concurrency in workflows can enhance efficiency and streamline processes, but it's crucial to ensure that pending workflows aren't inadvertently canceled. Proper management and coordination of concurrent tasks can prevent disruptions, allowing for a more seamless and productive workflow environment. Balancing concurrency with careful oversight ensures that all tasks are executed effectively without compromising existing processes.

**Reply by ghostinhershell** [reply]
Hi @paperwriter0300, we’re glad you’re here 😀

Thanks for sharing your website with the community. Unfortunately, we currently do not allow self-promotion or advertising on the Community on Discussions. We want to make sure there is space for users to ask questions without overwhelming them with other conversations. Thank you for helping us maintain a productive and tidy community for all our members.

**Comment by shinebayar-g** [comment]
I'm shook. This is not expected behavior from **concurrency** at all. Who in the right mind thought it was a good idea to remove queued items from the queue.. 🙄 
All we want is a basic FIFO queue. How is it even useful in its current state?

After months of migrating to _modern_ CI/CD platform, I can't believe we ended up here.

**Comment by wpride** [comment]
For anyone looking for a workaround, check out https://github.com/marketplace/actions/action-turnstyle

**Reply by Husnain08** [reply]
this does not work as intended. can you provide some examples @wpride ?

**Comment by farshad-t** [comment]
one workaround is to limit the workflow to only run on a single runner. This way, runner will limit the number of concurrent runs.

**Reply by taurit** [reply]
Wouldn't that only be possible with self-hosted runners?

I tried to explore that workaround option, but as far as I can understand, we cannot assign a label or create a single-runner group for GitHub-hosted runners, which would enable use of `runs-on: someLabelName` to achieve queueing.

**Reply by farshad-t** [reply]
It probably only works on a single self-hosted runner.

**Reply by Husnain08** [reply]
I am curious as to why it could work? can you explain it @farshad-t ?

**Reply by farshad-t** [reply]
If there is only one runner, there will be no need for "concurrency". Workflows will naturally run in order and one at a time.

**Reply by ephraim271** [reply]
@farshad-t I have only one self-hosted runner, but multiple workflows are still attempting to run at the same time

**Reply by vitalykarasik** [reply]
"attemping" - yes, i.e. we have a queue. And only one job is running, so it's a working workaround.

**Comment by GuillaumeDerivery** [comment]
+1 "cancel-pending: false"

**Comment by Kruziikrel13** [comment]
+1 "cancel-pending:false"

**Comment by Husnain08** [comment]
+1 "cancel-pending:false"

**Comment by Shereef** [comment]
I proposed a feature request here
https://github.com/orgs/community/discussions/12835
Please upvote it

**Reply by TSRBerry** [reply]
Oh I had no idea this existed! My post could have been just a comment to yours.
Thanks for sharing!

**Reply by Shereef** [reply]
@TSRBerry Are you able to put this link at the top ? I feel like we need to attract more people there

**Reply by TSRBerry** [reply]
Good idea, I'll do that!

**Comment by bcarranza** [comment]
Ey Github team, this is an important Lack , we really need this due requirements to work one workflow at once but without cancel , just enqueue.
do you have some information about feature.
+1 "cancel-pending:false"

?

**Comment by philn-delphia** [comment]
The behavior of concurrency groups is very surprising. This is a much needed feature.

**Comment by 9999years** [comment]


**Reply by aromeronavia** [reply]
So amazed concurrency doesn't work out of the box. I just got rid of the merge queue in my team because I thought it was not necessary, and all of a sudden, it seems we will need to go back to it because this feature is not complete...

**Comment by adamatsolita** [comment]
+1

**Comment by matheuscscp** [comment]
+1

**Comment by jmax01** [comment]
+1

**Comment by cmyllis** [comment]
+1. Would be funny if they implement it after 2 years 🥲

**Comment by tt-rkim** [comment]
👍  +1

**Comment by amy-brainlabs** [comment]
Hacky workaround we implemented, removing concurrency and adding these steps makes jobs sleep if any are already running and checks every 30 seconds:

```
jobs:
  build:
    runs-on: ubuntu-20.04
    steps:
      - name: Wait for previous runs to complete
        id: wait_for_previous_runs
        run: |
          IN_PROGRESS_RUNS=$(curl -s -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
                              -H "Accept: application/vnd.github.v3+json" \
                              "https://api.github.com/repos/${{ github.repository }}/actions/runs?event=push&status=in_progress" \
                              | jq -r '.workflow_runs[].id | select(. != ${{ github.run_id }})')
          if [ -z "$IN_PROGRESS_RUNS" ]; then
            echo "No previous runs in progress. Proceeding with deployment."
          else
            echo "Waiting for previous runs to complete..."
            echo "::set-output name=in_progress_runs::$IN_PROGRESS_RUNS"
          fi

      - name: Store in-progress runs
        if: steps.wait_for_previous_runs.outputs.in_progress_runs != ''
        id: store_in_progress_runs
        run: echo "::set-output name=in_progress_runs::${{ steps.wait_for_previous_runs.outputs.in_progress_runs }}"

      - name: Check if stored in-progress runs are completed
        if: steps.wait_for_previous_runs.outputs.in_progress_runs != ''
        id: check_in_progress_runs
        run: |
          while true; do
            STORED_RUNS="${{ steps.store_in_progress_runs.outputs.in_progress_runs }}"
            COMPLETED_RUNS=$(curl -s -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
                              -H "Accept: application/vnd.github.v3+json" \
                              "https://api.github.com/repos/${{ github.repository }}/actions/runs?event=push&status=completed" \
                              | jq -r '.workflow_runs[].id | select(. != ${{ github.run_id }})')
            if [ "$(comm -23 <(echo "$STORED_RUNS" | tr ' ' '\n' | sort) <(echo "$COMPLETED_RUNS" | tr ' ' '\n' | sort) | wc -l)" -eq 0 ]; then
              echo "All stored in-progress runs completed. Proceeding with deployment."
              break
            else
              echo "Waiting for stored in-progress runs to complete..."
              sleep 30
            fi
          done

          .... rest of steps
```

**Reply by Felixoid** [reply]
Sorry to say @amy-brainlabs, but it's a very limited approach.

Imagine a fleet of 600+ self-hosted runners sleeping for veeeeeerrrrrrrrrrry long time. It's just burning money for nothing

**Reply by amy-brainlabs** [reply]
yeah its works for us we have at most 4 or 5 short running processes, would be very costly for 600 and not appropriate!

**Reply by krukonshedul** [reply]
mine crypto while its waiting - just as wasteful, but spreads the profits 😅

**Reply by Anonymous-Coward** [reply]
> mine crypto while its waiting - just as wasteful, but spreads the profits 😅

I know this was meant as a joke, but, since I'm fun at parties anyway: you meant spread the costs. Mining crypto with unspecialized hardware is no longer profitable. If you do it on runners you pay for, you have to be excessively, unimaginably lucky to even break even. If you do it on runners you don't pay for, it's almost certainly illegal, and you;ll get caught.

**Reply by Anonymous-Coward** [reply]
I'm not completely confident in this mechanism. I can imagine situations in which a deadlock occurs. Also, it implies too many sleeping jobs - each one taking up one runner.

What I'd do instead is have one job acting as a controller, starting other jobs via manual dispatch, watching the number of running jobs and only starting additional jobs when there's space for them. You'd still have a job that mainly sleeps, but it would be just a single one.

**Reply by Felixoid** [reply]
So, writing a whole CI from the scratch, because of GHA.

I tend to agree with the approach. It doesn't look like the monopolist would like to do anything about broken parts.

**Reply by Anonymous-Coward** [reply]
> It doesn't look like the monopolist would like to do anything about broken parts.

There's always gitlab-ci, Jenkins X, Circle CI and a ton of others, if you don't like what GHA provides. Hardly any monopoly here.

**Reply by Felixoid** [reply]
The GH is defacto the default to store the code. Out of the box it provides half baked product for the CI.

Call it as you want, I'm just burn to the coal trying squeeze anything worthy from GHA.

**Comment by leonardo-kochake-gomes** [comment]
+1

**Comment by avdland** [comment]
+1

**Comment by archana-punia** [comment]
+1

**Comment by deynekas** [comment]
+1

**Comment by burns47** [comment]
+1

**Comment by Anonymous-Coward** [comment]
As far as I can tell, GHA provides the concurrency group feature for two very specific use cases:

1. when running one job would render the results of other jobs, that already ran, obsolete. Like when you run a build on master while a new PR is being merged into master. It makes no sense to finish the build if you need to start a new one right after that to incorporate the latest PR. 

2. When running a job must wait for another one to finish. Like when you run a library build and need to run an app build for an app using the library afterwards. It makes no sense to start building the app before the library build finishes, or you'd need to start the app build again, to incorporate the updated library.

Unfortunately, that's not the only use case where you'd want to control concurrency, or in only those two ways. Say your app builds are expensive, long-lasting workflows because besides the actual build they contain jobs that perform tons of complex validation and performance tests. And you don't have a single application that depends on the library, but many. You do want to perform all app builds, but don't want to run them in parallel, since that would either clog the entire CI for potentially hours, or make its scale explode, short term.

A solution, from the point of view of configuration, would be to have a concurrency section like this:

        concurrency:
            group: <string>
            strategy: cancel | cancel-running | queue
            max-concurrent: <number>

This would tell the controller that when jobs in the same concurrency group need to be executed, they should do exactly what they do now, if strategy is not `queue`, and just wait until fewer than <max-concurrent> jobs in the same concurrency group are running before starting. 

If multiple jobs in the same concurrency group provide different values for `max-concurrent`, the lowest value known at the time a job scheduling is slated to happen should be used and a warning issued in the annotations of all workflows that use the concurrency group, (plus a list of all workflows where the the same concurrency group is used).

**Comment by AshVz** [comment]
+1

**Comment by juho9000** [comment]
+1

**Comment by nebuk89** [comment]
I posted the same thing on the other feature request linked for this :) but also sharing here for completeness 
Sorry that the Actions team has been quiet on this. We have added this to the list of paper cuts we are hoping to address in the next 12 months :) I don't have a timeframe for this one, but I want to say we know it's a pain point and do have it on our list <3

**Comment by ivan-neverov-al** [comment]
+1

**Comment by heheksx** [comment]
+1

**Comment by Lawstorant** [comment]
We need this so bad. Currently, running terraform plans and applies needs constant attention. If someone posts a PR during infra apply, there goes rest of the components.

**Comment by shuozhao-work** [comment]
+1 we have a very busy pipeline being triggered very often, so we've got a lot of this type of errors.

**Comment by dmavrommatis** [comment]
+1

**Comment by ddusht** [comment]
+1

**Comment by nawanglama-krispcall** [comment]
+111

**Comment by DennisJensen95** [comment]
+1

**Comment by jjohnson1994** [comment]
+1

**Comment by KevinWorldGame** [comment]
+1

**Comment by ADavis11** [comment]
+1

**Comment by cpnielsen** [comment]
+1

**Comment by mracicot2** [comment]
+1

**Comment by leonardo-kochake-gomes** [comment]
This is required when using merge-queue.

The merge queue maximum parallel runs must be set to max 2, so it does not cancel intermediate runs.

It is far from optimal but as almost everything on Github.. it is a workaround.

**Comment by cars-matt-b** [comment]
+1

**Comment by juwon8891-sidewalk** [comment]
+1

**Comment by pgodinhofortis** [comment]
Getting close to the 2-year anniversary of this request! The first thing I do in the morning is pray that MSFT hires another developer to work on Actions, don't they realize the current skeleton crew can't push this product forward?

**Reply by leonardo-kochake-gomes** [reply]
I get the same feeling. 
Everywhere you look at the product, there is either a workaround or the community has created a custom action to mitigate a specific problem that should not have existed in the first place.

**Comment by ivan-guschin-rs** [comment]
+1 please fix the bug

**Comment by dorz-saymine** [comment]
+1

**Comment by astandrik** [comment]
+1

**Comment by voidgraphics** [comment]
+1

**Comment by bakavets-ma** [comment]
+1

**Comment by shink** [comment]
+1

**Comment by PhilGSilva** [comment]
+1

**Comment by eniasebiomo** [comment]
+1

**Comment by czhujer** [comment]
+1

**Comment by 0xangelo** [comment]
+1

**Comment by mattthaber** [comment]
+1

**Comment by itchyny** [comment]
No more +1 comment, use emoji reactions, please. I'm about to unsubscribe because of the meaningless notifications.

**Comment by malachany** [comment]
+1

**Comment by erickertz** [comment]
+1

**Comment by ui-kane-lin** [comment]
Push

**Comment by sudo-palash** [comment]
+1

**Comment by zdenko-kovac** [comment]
+1

**Comment by micmarty-deepsense** [comment]
+1

**Comment by robdunkley-kaplan** [comment]
+1 We need this

**Comment by Shereef** [comment, marked as answer]
I proposed a feature request here before this discussion and it's getting replies and being tracked by GHA team so please give all the attention there https://github.com/orgs/community/discussions/12835 Please upvote it

**Comment by nebuk89** [comment]
Oops sorry for the churn on the label, agree with @Shereef if we can keep this on 12835 please <3

**Comment by nadavbuc-orchid** [comment]
+1

**Comment by palionis** [comment]
+1

**Comment by Erikas-vainoras** [comment]
+1

**Comment by kristupasp** [comment]
+1

**Comment by brmb** [comment]
+1

**Comment by andriussilkinis** [comment]
+1

**Comment by andriusjurkus** [comment]
+1

**Comment by tauale** [comment]
+1

**Comment by RokasGad** [comment]
+1

**Comment by Homulvas** [comment]
Hey kids, could you maybe stop spamming everyone's notifications with the useless `+1`?

**Comment by Voveriukas** [comment]
+1

**Comment by 370rokas** [comment]
+1

**Comment by samus-aran** [comment]
Hey folks - I'm locking this thread to encourage folks to submit there interest by thumbs up reaction or upvote [via this discussion ](https://github.com/orgs/community/discussions/12835). 

<img width="620" height="148" alt="upvoting" src="https://github.com/user-attachments/assets/f863e3fd-a856-417b-9a7b-b5bb079b690e" />
